gpio/src/vhu_gpio.rs

```diff
@@ -0,0 +1,1189 @@
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        assert_eq!(
+            backend
+                .process_events(vec![desc_chain], &vring)
+                .unwrap_err(),
+            Error::UnexpectedDescriptorSize(size_of::<VirtioGpioIrqResponse>(), 100)
+        );
+
+        // Wait for event without setting irq type first
+        let flags: Vec<u16> = vec![0, VIRTQ_DESC_F_WRITE];
+        let len: Vec<u32> = vec![
+            size_of::<VirtioGpioIrqRequest>() as u32,
+            size_of::<VirtioGpioIrqResponse>() as u32,
+        ];
+        let desc_chain = prepare_desc_chain_dummy(None, flags, len);
+        let desc_chains = vec![desc_chain];
+        backend.process_events(desc_chains.clone(), &vring).unwrap();
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_IRQ_STATUS_INVALID, None);
+
+        // Wait for interrupt with failure msg
+        const GPIO: u16 = 5;
+        // Set irq type
+        let desc_chain = prepare_request_desc_chain(
+            GuestAddress(0),
+            VIRTIO_GPIO_MSG_IRQ_TYPE,
+            GPIO,
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH as u32,
+            1,
+        );
+        let desc_chains = vec![desc_chain];
+        backend
+            .process_requests(desc_chains.clone(), &vring)
+            .unwrap();
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_STATUS_OK, Some(vec![0]));
+
+        // Wait for interrupt
+        let desc_chain = prepare_event_desc_chain(GuestAddress(0), GPIO);
+        let desc_chains = vec![desc_chain];
+        backend.process_events(desc_chains.clone(), &vring).unwrap();
+
+        while backend.handles.read().unwrap()[GPIO as usize].is_some() {}
+        validate_desc_chains(desc_chains, VIRTIO_GPIO_IRQ_STATUS_INVALID, None);
+    }
+
+    #[test]
+    fn test_gpio_verify_backend() {
+        const NGPIO: u16 = 8;
+        let mut gpio_names = vec![
+            "gpio0".to_string(),
+            '\0'.to_string(),
+            "gpio2".to_string(),
+            '\0'.to_string(),
+            "gpio4".to_string(),
+            '\0'.to_string(),
+            "gpio6".to_string(),
+            '\0'.to_string(),
+        ];
+        // Controller adds '\0' for each line.
+        let names_size = std::mem::size_of_val(&gpio_names) + gpio_names.len();
+
+        let mut device = DummyDevice::new(NGPIO);
+        device.gpio_names.clear();
+        device.gpio_names.append(&mut gpio_names);
+        let controller = GpioController::new(device).unwrap();
+        let mut backend = VhostUserGpioBackend::new(controller).unwrap();
+
+        assert_eq!(backend.num_queues(), NUM_QUEUES);
+        assert_eq!(backend.max_queue_size(), QUEUE_SIZE);
+        assert_eq!(backend.features(), 0x171000001);
+        assert_eq!(
+            backend.protocol_features(),
+            VhostUserProtocolFeatures::MQ | VhostUserProtocolFeatures::CONFIG
+        );
+
+        assert_eq!(backend.queues_per_thread(), vec![0xffff_ffff]);
+
```

Comment:
Yeah, that's what is happening here, but now that you pointed it out, I am not sure if it adds any value. Though, since both the earlier implementations (i2c, rng) have it, maybe we can just add it.